src/array.rs

```diff
@@ -1485,12 +1331,857 @@ unsafe fn clone_elements<T: Element>(elems: &[T], data_ptr: &mut *mut T) {
+    /// For a non-panicking variant, use [`try_readonly`][Self::try_readonly].
+    fn readonly(&self) -> PyReadonlyArray<'py, T, D>
+    where
+        T: Element,
+        D: Dimension,
+    {
+        self.try_readonly().unwrap()
+    }
+
+    /// Get a mutable borrow of the NumPy array
+    fn try_readwrite(&self) -> Result<PyReadwriteArray<'py, T, D>, BorrowError>
+    where
+        T: Element,
+        D: Dimension;
+
+    /// Get a mutable borrow of the NumPy array
+    ///
+    /// # Panics
+    ///
+    /// Panics if the allocation backing the array is currently borrowed or
+    /// if the array is [flagged as][flags] not writeable.
+    ///
+    /// For a non-panicking variant, use [`try_readwrite`][Self::try_readwrite].
+    ///
+    /// [flags]: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flags.html
+    fn readwrite(&self) -> PyReadwriteArray<'py, T, D>
+    where
+        T: Element,
+        D: Dimension,
+    {
+        self.try_readwrite().unwrap()
+    }
+
+    /// Returns an [`ArrayView`] of the internal array.
+    ///
+    /// See also [`PyReadonlyArray::as_array`].
+    ///
+    /// # Safety
+    ///
+    /// Calling this method invalidates all exclusive references to the internal data, e.g. `&mut [T]` or `ArrayViewMut`.
+    unsafe fn as_array(&self) -> ArrayView<'_, T, D>
+    where
+        T: Element,
+        D: Dimension;
+
+    /// Returns an [`ArrayViewMut`] of the internal array.
+    ///
+    /// See also [`PyReadwriteArray::as_array_mut`].
+    ///
+    /// # Safety
+    ///
+    /// Calling this method invalidates all other references to the internal data, e.g. `ArrayView` or `ArrayViewMut`.
+    unsafe fn as_array_mut(&self) -> ArrayViewMut<'_, T, D>
+    where
+        T: Element,
+        D: Dimension;
+
+    /// Returns the internal array as [`RawArrayView`] enabling element access via raw pointers
+    fn as_raw_array(&self) -> RawArrayView<T, D>
+    where
+        T: Element,
+        D: Dimension;
+
+    /// Returns the internal array as [`RawArrayViewMut`] enabling element access via raw pointers
+    fn as_raw_array_mut(&self) -> RawArrayViewMut<T, D>
+    where
+        T: Element,
+        D: Dimension;
+
+    /// Get a copy of the array as an [`ndarray::Array`].
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use numpy::PyArray;
+    /// use ndarray::array;
+    /// use pyo3::Python;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let pyarray = PyArray::arange(py, 0, 4, 1).reshape([2, 2]).unwrap();
+    ///
+    ///     assert_eq!(
+    ///         pyarray.to_owned_array(),
+    ///         array![[0, 1], [2, 3]]
+    ///     )
+    /// });
+    /// ```
+    fn to_owned_array(&self) -> Array<T, D>
+    where
+        T: Element,
+        D: Dimension,
+    {
+        unsafe { self.as_array() }.to_owned()
+    }
+
+    /// Get the single element of a zero-dimensional array.
+    ///
+    /// See [`inner`][crate::inner] for an example.
+    fn item(&self) -> T
```

Comment:
Something is off here: This method should only exist for zero dimensional arrays (note that its impl block has the `D` parameter filled in). I suspect we need a separate trait like `PyArray0Methods` for this?

A similar reduction of parametricity also apply to various one-, two- and three-dimensional constructors like `from_slice` which should probably get `_bound` variants? And also `from_owned_object_array` which however constains the element type instead of the dimension.

Reply:
> Something is off here: This method should only exist for zero dimensional arrays (note that its impl block has the `D` parameter filled in). I suspect we need a separate trait like `PyArray0Methods` for this?

True, I must have overlooked that. I will adjust accordingly.

> A similar reduction of parametricity also apply to various one-, two- and three-dimensional constructors like `from_slice` which should probably get `_bound` variants?

I have split off constructors for later PRs to keep the size of this manageable.

> And also `from_owned_object_array` which however constrains the element type instead of the dimension.

I'll have a look